### ISSUE TITLE
Server Side Rendering and ReactJS.NET

### ISSUE THREAD

**Issue by daveaspinall** [styled-components/styled-components#1290]
We're currently working on a project which uses ReactJS.NET to render our React components server side, and then our client bundle takes over on the frontend once loaded.

Our styled components work perfectly client side, but we're unable to get them working server side. All the examples we can see create a stylesheet using `ServerStyleSheet` and then render the results to html...but we can't do this in the .NET environment. Is there another way?

We tried the following which has some success on the server, but when the client takes over we lose all styles (and we're not actually sure why it's working since we don't run `collectStyles` on the `sheet`):

```
// ServerApp.js

const generateStyles = () => {
  const sheet = new ServerStyleSheet()
  return sheet.getStyleElement()
}

const ServerApp = (props) => {
  const styleTags = generateStyles()
  
  return (
    <Provider store={configureStore(props.initialState)}>
      <StaticRouter location={props.path} context={props.context}>
        {styleTags}
        <App />
      </StaticRouter>
    </Provider>
  )
}
```

The above only works when we include the following in our Babel config (but we lose styles when the js kicks in):

```
{
  "plugins": [
    ["babel-plugin-styled-components", {
      "ssr": true
    }]
  ]
}
```

Are there any examples of using Styled Components with ReactJS.NET server side, or it is not even possible without `renderToHtml`?

Thanks!

**Comment by philpl** [comment]
@daveaspinall I’m not sure how you actually call into React from .NET, but our server Stylesheet is not this simple.

It’s connected to every styled component like a provider via the context. Then all styling is injected into the server style sheet instead and those are the styles we collect.

So one way or another you do have to run collectStyles or use the StyleSheetManager, get the style HTML or tags, and put them into the servers head result.

Otherwise it really won’t work. Maybe if you tell me what you can do, i can try to come up with a workaround

**Comment by daveaspinall** [comment]
@philpl thanks for the help!

We're using React.NET to render an embedded React app/component in a partial view in an ASP.NET MVC application, because of this we don't have full control of the response, i.e. we can't manipulate the <head> tag from where our app/component is rendered.

Essentially we create a React component, export it as a global variable in our JS bundle, and just hand that over to .NET to render with data from the server.

We can use `collectStyles` to output a `<style>` *somewhere* in the DOM but we can't put it in the head tag with SSR. This works fine when JavaScript is disabled, we see a response where our components are fully styled with a `<style>` tag inserted just above where our component renders. However with JavaScript enabled, when the client bundle takes over the `<style>` tag that is inserted server side is removed and an empty `<style>` tag is added to the head tag.

In summary:

1. We can use `collectStyles` server side to output a `<style>` tag in the DOM, but not in the `<head>` tag, this works for the initial render.
2. When the client bundle takes over the `<style>` tag that is initially rendered is removed and an empty `<style>` tag is added to the `<head>` tag.
3. If we don't output a `<style>` tag server side, the client works as expected and adds a fully populated `<style>` tag to `<head>`.

Hope that helps give you a better understanding! Happy to provide more info if needed 🙂

**Comment by philpl** [comment]
I think you’ll need to wait on this PR then: https://github.com/styled-components/styled-components/pull/1102

Unfortunately none of us have been actively addressing its problems yet 😔

**Comment by daveaspinall** [comment]
@philpl that solution would be perfect I think! Typical we need it just before it's probably due to be worked on 😄

If you guys need any help at all let me know. We've got two projects on right now that will need this, for some large clients, and I dread going back to normal stylesheets.

**Comment by daveaspinall** [comment]
@philpl okay, a quick update! We've managed to get our components working both server-side and client-side by forcing Styled Components to always generate the client-side style tag.

Essentially it was checking to see if the server-side `<style>` tag had been created, it found one in the DOM so didn't create a new one, then the `<style>` tag inside the page content is removed so we're left with no styles.

It's hacky but we got it working by adding this on line 152 of `src/models/BrowserStyleSheet.js`:

```
tags = []
names = {}
```

Obviously it's not optimal but it gets around our problem until #1102 gets resolved.

**Comment by dustinsoftware** [comment]
@daveaspinall we have implemented full support for server-rendered styles, check it out here:

https://reactjs.net/features/css-in-js.html

**Comment by mxstbr** [comment]
Amazing @dustinsoftware, I guess this can be closed! :tada:
